File: testclient.c
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>

int main() {
	int fd = open("/dev/parrot", O_RDONLY);
	if (fd < 0) {
		printf("%s\n", strerror(-fd));
		return -fd;
	}

	char buf[2];
	while (1) {
		// Buffer needs to be manually null terminated.
		int r = read(fd, buf, 1);
		if (r < 0) {
			printf("%s\n", strerror(-r));
			return -r;
		} else {
			buf[r] = 0;
		}
		printf("%s", buf);
	}
}

File: testclient.rs
use std::{error::Error, fs::File, io::Read, str::from_utf8};

fn main() -> Result<(), Box<dyn Error>> {
    let mut f = File::open("/dev/parrot")?;
    let buffer = &mut [0; 4096];
    loop {
        let read = f.read(buffer)?;
        print!("{}", from_utf8(&buffer[..read])?);
    }
}
